File: src/molecule.cpp
#include "molecule.h"

#include "element_utils.h"

#include <fstream>
#include <sstream>
#include <iostream>

Molecule::AtomID Molecule::add_atom(const std::string& element, double x, double y, double z) {
    int atomic_number = get_atomic_number(element);
    atoms.push_back({num_atoms, element, atomic_number, x, y, z});
    adjacency[num_atoms] = {};
    return num_atoms++;
}

void Molecule::add_bond(AtomID a1, AtomID a2, int order) {
    bonds.push_back({a1, a2, order});
    adjacency[a1].push_back(a2);
    adjacency[a2].push_back(a1);
}

const std::vector<Molecule::Atom>& Molecule::get_atoms() const {
    return atoms;
}

const std::vector<Molecule::Bond>& Molecule::get_bonds() const {
    return bonds;
}

std::vector<Molecule::AtomID> Molecule::neighbors(AtomID id) const {
    auto it = adjacency.find(id);
    if (it != adjacency.end()) {
        return it->second;
    }
    return {};
}

bool Molecule::empty() const {
    return atoms.empty();
}

std::vector<Molecule> load_from_mol2(const std::string& filename) {
    std::ifstream infile(filename);
    if (!infile) {
        throw std::runtime_error("File not found");
    }

    std::vector<Molecule> molecules;
    Molecule current;

    std::string line;
    bool in_atom_section = false;
    bool in_bond_section = false;

    while (std::getline(infile, line)) {
        if (line.find("@<TRIPOS>MOLECULE") != std::string::npos) {
            if (!current.empty()) {
                molecules.push_back(current);
                current = Molecule();  // Start fresh
            }
            in_atom_section = in_bond_section = false;
        } else if (line.find("@<TRIPOS>ATOM") != std::string::npos) {
            in_atom_section = true;
            in_bond_section = false;
            continue;
        } else if (line.find("@<TRIPOS>BOND") != std::string::npos) {
            in_atom_section = false;
            in_bond_section = true;
            continue;
        } else if (line.size() && line[0] == '@') {
            in_atom_section = in_bond_section = false;
            continue;
        }

        if (in_atom_section) {
            std::istringstream ss(line);
            int id;
            std::string name, element;
            double x, y, z;
            ss >> id >> name >> x >> y >> z >> element;
            current.add_atom(element, x, y, z);
        } else if (in_bond_section) {
            std::istringstream ss(line);
            int bond_id, a1, a2;
            std::string order_str;
            ss >> bond_id >> a1 >> a2 >> order_str;

            int order = 1;
            if (order_str == "2") order = 2;
            else if (order_str == "3") order = 3;
            else if (order_str == "ar") order = 1; // aromatic

            current.add_bond(a1 - 1, a2 - 1, order);
        }
    }

    // Add the last molecule if it had content
    if (!current.empty()) {
        molecules.push_back(current);
    }

    return molecules;
}

std::vector<Molecule> filter_molecules(const std::vector<Molecule>& mols)
{
    // Apply 3 filters to the list of molecules:
    // 1. Only molecules with atoms <= 25 are allowed.
    // 2. Only molecules with ALL single bonds are allowed.
    // 3. Only molecules with more than 1 HBA and 1 HBD are allowed.
}

int Molecule::countHBondAcceptors() const {
    int count = 0;
    // Complete this function!
    return count;
}

int Molecule::countHBondDonors() const {
    int count = 0;
    // Complete this function!
    return count;
}

// Helper function
int get_bond_order(Molecule::AtomID a1, Molecule::AtomID a2, std::vector<Molecule::Bond> bonds)
{
    for (auto& bond: bonds) {
        if (bond.atom1 == a1 && bond.atom2 == a2) {
            return bond.order;
        }
        if (bond.atom1 == a2 && bond.atom2 == a1) {
            return bond.order;
        }
    }
    return 0;
}

File: molecule/src/molecule.rs
use std::collections::HashMap;

type AtomID = usize;

#[derive(Debug, Clone)]
pub struct Atom {
    pub id: AtomID,
    pub element: String,
    pub atomic_number: i32,
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

#[derive(Debug, Clone)]
pub struct Bond {
    pub atom1: AtomID,
    pub atom2: AtomID,
    pub order: i32,
}

pub struct Molecule {
    atoms: Vec<Atom>,
    bonds: Vec<Bond>,
    adjacency: HashMap<AtomID, Vec<AtomID>>,
    num_atoms: AtomID,
}

impl Molecule {
    pub fn new() -> Self {
        Self {
            atoms: Vec::new(),
            bonds: Vec::new(),
            adjacency: HashMap::new(),
            num_atoms: 0,
        }
    }

    pub fn add_atom(&mut self, element: &str, x: f64, y: f64, z: f64) -> AtomID {
        let atomic_number = get_atomic_number(element);
        let id = self.num_atoms;
        self.atoms.push(Atom {
            id,
            element: element.to_string(),
            atomic_number,
            x,
            y,
            z,
        });
        self.adjacency.insert(id, Vec::new());
        self.num_atoms += 1;
        id
    }

    pub fn add_bond(&mut self, a1: AtomID, a2: AtomID, order: i32) {
        self.bonds.push(Bond { atom1: a1, atom2: a2, order });
        self.adjacency.entry(a1).or_default().push(a2);
        self.adjacency.entry(a2).or_default().push(a1);
    }

    pub fn get_atoms(&self) -> &Vec<Atom> {
        &self.atoms
    }

    pub fn get_bonds(&self) -> &Vec<Bond> {
        &self.bonds
    }

    pub fn neighbors(&self, id: AtomID) -> Vec<AtomID> {
        self.adjacency.get(&id).cloned().unwrap_or_default()
    }

    pub fn empty(&self) -> bool {
        self.atoms.is_empty()
    }

    pub fn count_hbond_acceptors(&self) -> i32 {
        let mut count = 0;

        for atom in &self.atoms {
            let neighbor_ids = self.neighbors(atom.id);
            let total_bond_order: i32 = neighbor_ids
                .iter()
                .map(|&nid| get_bond_order(atom.id, nid, &self.bonds))
                .sum();

            if atom.element == "N" && total_bond_order == 3 {
                count += 1;
            }
            if atom.element == "O" && total_bond_order == 2 {
                count += 1;
            }
        }

        count
    }

    pub fn count_hbond_donors(&self) -> i32 {
        let mut count = 0;

        for atom in &self.atoms {
            let neighbor_ids = self.neighbors(atom.id);
            let total_bond_order: i32 = neighbor_ids
                .iter()
                .map(|&nid| get_bond_order(atom.id, nid, &self.bonds))
                .sum();

            if atom.element == "N" && total_bond_order == 3 {
                if neighbor_ids.iter().any(|&nid| self.atoms[nid].element == "H") {
                    count += 1;
                }
            }

            if atom.element == "O" && total_bond_order == 2 {
                if neighbor_ids.iter().any(|&nid| self.atoms[nid].element == "H") {
                    count += 1;
                }
            }
        }

        count
    }
}

fn get_atomic_number(element: &str) -> i32 {
    match element {
        "H" => 1,
        "C" => 6,
        "N" => 7,
        "O" => 8,
        _ => 0,
    }
}

/// Helper function to determine bond order between two atoms
pub fn get_bond_order(a1: AtomID, a2: AtomID, bonds: &[Bond]) -> i32 {
    for bond in bonds {
        if (bond.atom1 == a1 && bond.atom2 == a2) || (bond.atom1 == a2 && bond.atom2 == a1) {
            return bond.order;
        }
    }
    0
}
